### PR TITLE
[MIG] 8.0 hr_timesheet_improvement

### PR DESCRIPTION
**Pull request by charbeljc**

### CODE REVIEW

**Comment by vrenaville** [comment]
:+1: code review + test

**Comment by lepistone** [comment]
@charbeljc can you please check if it is still WIP? Thanks!

**Comment by charbeljc** [comment]
@lepistone I didn't done an extensive test of this module, but I gues I could remove the WIP status.

File: hr_timesheet_improvement/hr_attendance.py [hunk 1]
```diff
@@ -38,9 +38,10 @@ def _default_date(self, cr, uid, context=None):
         if not dates:
             return timesheet.date_from
         return max(dates)
-    """
-    Check sign in signout in the same timesheet only
-    """
+
+    # HUM CHECKME
```
Comment (pedrobaeza) [inline hr_timesheet_improvement/hr_attendance.py]: This comment in between is not recommended. Why is here?

**Comment by pedrobaeza** [comment]
Please extract module description to a README.rst file with OCA's template.

**Comment by mdietrichc2c** [comment]
It would be worth it to migrate `hr_timesheet_improvement/hr_timesheet.py` to the new API (models.Model, new fields, etc)

File: hr_timesheet_improvement/README.rst [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+Credits
+=======
+
+Contributors
+------------
+
+* Vincent Renaville <vincent.renaville@camptocamp.com>
+* Alexandre Fayolle <alexandre.fayolle@camptocamp.com>
+* Yannick Vaucher <yannick.vaucher@camptocamp.com>
+* Pedro M. Baeza <pedro.baeza@gmail.com>
```
Comment (yvaucher) [inline hr_timesheet_improvement/README.rst]: Not sure we want to list @pedrobaeza and @gurneyalex as contributors for PEP8 formating:

https://github.com/OCA/hr-timesheet/commit/b14a2bff850991518acfaee720a614ebda2ea8c2#diff-02c5bf21eb5a8e0c08f11b8c8b7c1bdb
https://github.com/OCA/hr-timesheet/commit/1b1232dd451cce6a4af9bbe814ed9ffd50ca8606#diff-02c5bf21eb5a8e0c08f11b8c8b7c1bdb

Personally I don't want my name in modules I only have performed PEP8 fixes.

@gurneyalex and @pedrobaeza Your opinion is welcome on this. Especially, do you want to be a reference contact for this module?

**Comment by yvaucher** [comment]
Just a little question about authors. Otherwise :+1: 

It's great to have some tests in there :smile:

File: hr_timesheet_improvement/README.rst (hunk 2 above)
Comment (pedrobaeza) [inline hr_timesheet_improvement/README.rst]: Indeed my work was very little in this module, so technically I'm a contributor, but doesn't mind if I appear or not. It's up to you.

**Comment by mdietrichc2c** [comment]
:+1: thanks for migrating to the new API

File: hr_timesheet_improvement/tests/test_timesheet_improvement.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        sheet_obj = self.env['hr_timesheet_sheet.sheet']
+        sheet = sheet_obj.create({'user_id': self.user.id,
+                                  'employee_id': self.employee.id,
+                                  'date_from': fields.Date.today()})
+        att_rs = self.attendance_model.with_context(sheet_id=sheet.id).create(
+            {'action': 'sign_in',
+             'employee_id': self.employee.id,
+             }
+            )
+        att_name = datetime.datetime.strptime(att_rs.name,
```
Comment (lmignon) [inline hr_timesheet_improvement/tests/test_timesheet_improvement.py]:
``` python
att_name = fields.Datetime.from_string(att_rs.name)
```

File: hr_timesheet_improvement/tests/test_timesheet_improvement.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                                              DEFAULT_SERVER_DATETIME_FORMAT)
+        att_name = att_name.strftime(DEFAULT_SERVER_DATE_FORMAT)
```
Comment (lmignon) [inline hr_timesheet_improvement/tests/test_timesheet_improvement.py]: att_name = fields.Date.to_string(att_name)

File: hr_timesheet_improvement/tests/test_timesheet_improvement.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+            self.attendance_model.with_context(sheet_id=s.id)._default_date()
+            )
+        self.assertEqual(old_date, att_rs.name)
+        new_date = (
+            fields.Datetime.from_string(old_date) + datetime.timedelta(hours=1)
+            )
+        att_rs1 = self.attendance_model.with_context(sheet_id=s.id).create(
+            {'action': 'sign_out',
+             'employee_id': self.employee.id,
+             'name': new_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
```
Comment (lmignon) [inline hr_timesheet_improvement/tests/test_timesheet_improvement.py]:
``` python
'name' : fields.Datetime.to_string(new_date)
```

File: hr_timesheet_improvement/tests/test_timesheet_improvement.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        self.assertEqual(old_date, att_rs1.name)
+
+    def test_timesheet_without_existing_timesheet(self):
+        att_rs = self.attendance_model.create(
+            {'action': 'sign_in',
+             'employee_id': self.employee.id,
+             }
+            )
+        att_name = fields.Datetime.from_string(att_rs.name)
+        att_name = att_name.strftime(DEFAULT_SERVER_DATE_FORMAT)
```
Comment (lmignon) [inline hr_timesheet_improvement/tests/test_timesheet_improvement.py]:
``` python
att_name = fields.Date.to_string(att_name)
```

File: hr_timesheet_improvement/tests/test_timesheet_improvement.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            self.attendance_model.with_context(sheet_id=s.id)._default_date()
+            )
+        new_date = (
+            fields.Datetime.from_string(old_date) + datetime.timedelta(hours=1)
+            )
+        with self.assertRaises(exceptions.ValidationError):
+            self.attendance_model.with_context(sheet_id=s.id).create(
+                {'action': 'sign_in',
+                 'employee_id': self.employee.id,
+                 'name': new_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
```
Comment (lmignon) [inline hr_timesheet_improvement/tests/test_timesheet_improvement.py]:
``` python
'name' : fields.Datetime.to_string(new_date)
```

File: hr_timesheet_improvement/README.rst (hunk 2 above)
Comment (yvaucher) [inline hr_timesheet_improvement/README.rst]: @pedrobaeza My point is that doing meta contribution like a bit of PEP8 on all modules or adding a field in `__openerp__.py` would show some names everywhere. :smile:

File: hr_timesheet_improvement/tests/test_timesheet_improvement.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+
+import datetime
+import time
+from openerp import fields, exceptions
+from openerp.tests import common
+from openerp.tools import DEFAULT_SERVER_DATETIME_FORMAT
+from openerp.tools import DEFAULT_SERVER_DATE_FORMAT
```
Comment (lmignon) [inline hr_timesheet_improvement/tests/test_timesheet_improvement.py]: You should remove these two last import and use functions `from_string` and `to_string` defined on on classes `fields.Date` and `fields.Datetime` to convert your datetime objects in the from/to the format expected by the server.

**Comment by damdam-s** [comment]
fixes done

File: hr_timesheet_improvement/README.rst [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+Timesheet improvements
+======================
+
+Modifies timesheet behavior:
+ * Ensure a DESC order on timesheet lines
+ * Set default date for manually entering attendance to max attendance date
+ * Redefine constraint on timesheets to check alternation of 'sign in' and
+   'sign out' only on current timesheet instead of doing it on all timesheets
+   of the employee
+
```
Comment (lmignon) [inline hr_timesheet_improvement/README.rst:13]: Can you add the 'Usage' section with the new 'Try me on runbot' button https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst#usage

**Comment by charbeljc** [comment]
Thanks @damdam-s :+1:

**Comment by lmignon** [comment]
@damdam-s Thank you for this good job! It would be nice to have the new button 'try me on runbot' into the description otherwise :+1:

**Comment by damdam-s** [comment]
@lmignon runbot button added. I didn't know about that

**Comment by lmignon** [comment]
@damdam-s It's new :smirk: added in the sample template 7 days ago.
I count three thumbs up. @pedrobaeza this one is ready to merge.

**Comment by pedrobaeza** [comment]
OK, merging.

**Comment by lmignon** [comment]
Thank you @pedrobaeza

**Comment by rafaelbn** [comment]
Please @charbeljc , I don't understand usage of this module. I have try to find improvement in runbot,
I have compared with Odoo's runbot. In my opinion it's not clear.

> ```
>    Ensure a DESC order on timesheet lines
>    Set default date for manually entering attendance to max attendance date
>    Redefine constraint on timesheets to check alternation of 'sign in' and 'sign out' only on current timesheet instead of doing it on all timesheets of the employee
> ```

Think in people who do not read code please. Think in Odoo Apps: https://www.odoo.com/apps

For example for trying this un runbot

> ```
>    Set default date for manually entering attendance to max attendance date
> ```

You need to go to users - Technifal features an select (activate) "attendance" for the admin user...

Let me know you opinion

Thanks

**Comment by lepistone** [comment]
@charbeljc or @damdam-s, can you please make a new PR to fix the authorship of the module? Thanks a lot!

**Comment by lepistone** [comment]
As for @rafaelbn's remark, It sounds reasonable to me but less severe than wrong attribution. So I guess the documentation will have to be improved in a new PR if someone can contribute it. Thanks to all!

**Comment by damdam-s** [comment]
@lepistone @yvaucher done in PR https://github.com/OCA/hr-timesheet/pull/55

**Pull request merged**